### PR TITLE
Added upvotes to the backend

### PR DESCRIPTION
**Pull request by dschaadt32**
Two new routes,   /api/upvotes/{requestId} to post and /api/upvotes/upvotes?requestId=requestId to delete. It mutates the upvotingUsers array under the request object and saves it to the database and returns the requestResponse.

### CODE REVIEW

**Comment by jasyyang** [comment]
1) i rebased with main so that the backend + docs for Space reflects whatever is on prod right now and everything deploys.
2) [98328aa](https://github.com/camtheman256/6.1040-final-project/pull/47/commits/98328aafa31436ce3487d3ba282041079eff1ac0) ensures that we have populated updatedUsers in PlaceRequest responses; just reflects [this](https://github.com/camtheman256/6.1040-final-project/issues/42#issuecomment-1345632015)
3) [3c53a47](https://github.com/camtheman256/6.1040-final-project/pull/47/commits/3c53a47fe481a2b147f238949bbed872d776b4ab): had to pull in the mongo id for the session User because we use gapiUserIds for sessions. (prolly should've used mongo_id to begin with knowing how often we need to populate users but this works)

**Review by jasyyang — APPROVED** [review]
might be worth doing some light testing but otherwise looks good to me!

**Comment by jasyyang** [comment]
wait, we may need to call `request.populate("upvotingUsers")` wherever you return a new request response in `/upvote/router.ts`

File: backend/space/middleware.ts [hunk 1]
```diff
@@ -42,7 +42,7 @@ const isPlaceAlreadyExists = async (req: Request, res: Response, next: NextFunct
  * Checks if place_id in req.params exists
  */
 const isPlaceExists = async (req: Request, res: Response, next: NextFunction) => {
-    const place_id: string = req.params.place_id
```
Comment (jas-yang) [inline backend/space/middleware.ts:45]: i believe this should still be params. may need to ignore changes on this branch for `backend/space/middleware.ts` for merge

File: backend/upvote/router.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+      return;
+    };
+    if (request.upvotingUsers.includes(sessionUser?._id)) {
+      // this should not happen
+      res.status(201).json({
+        message: "User has already upvoted this response",
+        request: request,
+      });
+    } else {
+      request.upvotingUsers.push(sessionUser?._id);
```
Comment (jasyyang) [inline backend/upvote/router.ts:48]: after this push, we need a way to repopulate `request.upvotingUsers`

Possibly we can move the check to see if the user already upvoted the request to `/request/middleware.ts` and add a function in `/request/collection.ts` like ```updateOne(requestId: string, upvotingUser: string): Promise<HydratedDocument<Request>>``` to add an "upvoter" then return the populated request object.

File: backend/upvote/router.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      res.status(500).json({ //should never reach here
+        message: "Session user cannot be found"
+      });
+      return;
+    };
+    if (!request.upvotingUsers.includes(sessionUser._id)) {
+      // this should not happen
+      res.status(201).json({
+        message: "User did not upvote this response",
+        request: request,
```
Comment (jasyyang) [inline backend/upvote/router.ts:91]: also needs to be repopulated

**Review by jasyyang — CHANGES_REQUESTED** [review]
need to repopulate requests

**Pull request merged**